services/horizon/internal/handler.go

```diff
@@ -363,17 +363,27 @@ type pageAction interface {
 	GetResourcePage(w actions.HeaderWriter, r *http.Request) ([]hal.Pageable, error)
 }
 
+type buildPageFun func(r *http.Request, records []hal.Pageable) (hal.Page, error)
+
 type pageActionHandler struct {
 	action         pageAction
 	streamable     bool
 	streamHandler  sse.StreamHandler
 	repeatableRead bool
+	buildPage      buildPageFun
 }
 
 func restPageHandler(action pageAction) pageActionHandler {
 	return pageActionHandler{action: action}
 }
 
+func restPageHandlerWithBuilder(action pageAction, customPageBuilder buildPageFun) pageActionHandler {
+	return pageActionHandler{
+		action:    action,
+		buildPage: customPageBuilder,
+	}
+}
```

Comment:
nitpick: I think it's nicer if you use an interface instead of a function

```suggestion
type pageActionHandler struct {
	action         pageAction
	streamable     bool
	streamHandler  sse.StreamHandler
	repeatableRead bool
	buildPage      buildPageFun
}

func restPageHandler(action pageAction) pageActionHandler {
	return pageActionHandler{action: action}
}

interface customBuiltPageAction {
	pageAction
	BuildPage (r *http.Request, records []hal.Pageable) (hal.Page, error)
}

func restCustomBuiltPageHandler(action customBuiltPageAction) pageActionHandler {
	return pageActionHandler{
		action:    action,
		buildPage: customBuiltPageAction.BuildPage,
	}
}
```